src/blocks/ExperienceRoot.tsx

```diff
@@ -1,22 +1,44 @@
-import React from 'react'
-import { Experience } from '../types'
+import React, { useCallback, useEffect, useState } from 'react'
+import { DeprecatedExperience, Experience, InternalSDKMode } from '../types'
 import { VisualEditorRoot } from './VisualEditorRoot'
 import { PreviewDeliveryRoot } from './PreviewDeliveryRoot'
 import { supportedModes } from '../constants'
 import { validateExperienceBuilderConfig } from '../validation'
 import { ErrorBoundary } from './ErrorBoundary'
+import { isDeprecatedExperience } from '../typeguards'
+import { DeprecatedPreviewDeliveryRoot } from './DeprecatedPreviewDeliveryRoot'
 
 type ExperienceRootProps = {
-  experience: Experience
+  experience: Experience | DeprecatedExperience
   locale: string
-  slug: string
+  /**
+   * @deprecated
+   */
+  slug?: string
 }
 
 export const ExperienceRoot = ({ locale, experience, slug }: ExperienceRootProps) => {
-  const { mode } = experience
+  const [mode, setMode] = useState<InternalSDKMode>(() => {
+    if (supportedModes.includes(experience.mode)) {
+      return experience.mode
+    }
+
+    throw new Error(
+      `Unsupported mode provided: ${experience.mode}. Supported values: ${supportedModes}`
+    )
+  })
+
+  useEffect(() => {
+    if (supportedModes.includes(mode)) {
+      setMode(mode)
+    }
+  }, [mode])
```

Comment:
Can you elaborate what this is and why we need this rendering-based validation of the state variable? As far as I see, `setMode` is only called once with a hardcoded string, so this condition will never be true. So maybe, we can remove this whole effect atm?

Reply:
Before this change, this useEffect was living in `useExperienceBuilder` hook, which handled the switch between the modes, but after the changes in this PR, I moved it into `ExperienceRoot`. 

I think we will be able to remove it in the next bump of the major version

Reply:
Not a blocker as it doesn't do anything but I still don't get why we need to keep it as it doesn't do anything.
Or do you rather mean that this is just a moved code and thus not part of this PR to clean this up? (which is a reasonable argument)